components/camera/replaypcd/replaypcd_utils_test.go

```diff
@@ -0,0 +1,193 @@
+	}
+
+	r := &inject.Robot{}
+	rs := map[resource.Name]resource.Resource{}
+	rs[cloud.InternalServiceName] = mockCloudConnectionService
+	r.MockResourcesFromMap(rs)
+
+	return resourcesFromDeps(t, r, []string{cloud.InternalServiceName.String()}), rpcServer.Stop
+}
+
+// createNewReplayPCDCamera will create a new replay_pcd camera based on the provided config with either
+// a valid or invalid data client.
+func createNewReplayPCDCamera(ctx context.Context, t *testing.T, replayCamCfg *Config, validDeps bool,
+) (camera.Camera, func() error, error) {
+	logger := golog.NewTestLogger(t)
+
+	// var resources resource.Dependencies
+	// var closeRPCFunc func() error
+	// if validDeps {
+	resources, closeRPCFunc := createMockCloudDependencies(ctx, t, logger, validDeps)
+	// }
+
+	cfg := resource.Config{ConvertedAttributes: replayCamCfg}
+	cam, err := newPCDCamera(ctx, resources, cfg, logger)
+
+	return cam, closeRPCFunc, err
+}
+
+// resourcesFromDeps returns a list of dependencies from the provided robot.
+func resourcesFromDeps(t *testing.T, r robot.Robot, deps []string) resource.Dependencies {
+	t.Helper()
+	resources := resource.Dependencies{}
+	for _, dep := range deps {
+		resName, err := resource.NewFromString(dep)
+		test.That(t, err, test.ShouldBeNil)
+		res, err := r.ResourceByName(resName)
+		if err == nil {
+			// some resources are weakly linked
+			resources[resName] = res
+		}
+	}
+	return resources
+}
+
+// getNextDataAfterFilter returns the next point cloud data to be return based on the provided filter and
+// last returned item.
+func getNextDataAfterFilter(filter *datapb.Filter, last string) (int, error) {
+	// Apply the time-based filter based on the second's value in the start and end fields. Because artifacts
+	// do not have timestamps associated with them but are numerically ordered we can approximate the filtering
+	// by sorting for the files which occur after the start second count and before the end second count.
+	start := 0
+	end := 1000
+	startTime := filter.Interval.Start
+	if startTime != nil {
+		start = startTime.AsTime().Second()
+	}
+	endTime := filter.Interval.End
+	if endTime != nil {
+		end = endTime.AsTime().Second()
+	}
+
+	possibleData := makeFilteredRange(0, numPCDFiles, start, end)
+
+	// Return most recent file after last (should it exist)
+	if last == "" {
+		if len(possibleData) != 0 {
+			return possibleData[0], nil
+		}
+	} else {
+		lastFileNum, err := strconv.Atoi(last)
+		if err != nil {
+			return 0, err
+		}
+		for i := range possibleData {
+			if possibleData[i] > lastFileNum {
+				return possibleData[i], nil
+			}
+		}
+	}
+	return 0, errEndOfDataset
+}
+
+// makeFilteredRange returns a numerical range of values representing the possible files to be returned
+// after filtering.
+func makeFilteredRange(min, max, start, end int) []int {
+	a := []int{}
+	for i := 0; i < max-min; i++ {
+		val := min + i
+		if val >= start && val < end {
+			a = append(a, val)
+		}
+	}
+	return a
```

Comment:
This logic feels a bit unnecessarily complicated to me, with getting the range of files and then iterating over it. Can't we just do something like:

```
startTime := -1
endTime := -1
if startTime != nil { startTime = startTime.AsTime().Second() }
if endTime != nil { endTime = endTime.AsTime().Second() }
if startTime != -1 && startTime > numPCDFiles { return nil }
if endTime == -1 || endTime <= numPCDFiles { 
    if !last {
        return getFile(endTime)
    }
    if last >= numPCDFiles {
        return nil
    }
    return getFile(last)
}
```
This is pseudocode but hopefully makes sense - happy to sync up offline.